kv_test.go

```diff
@@ -195,6 +196,123 @@ func TestCompareAndDelete(t *testing.T) {
 	assert.True(t, deleted)
 }
 
+func TestSetAtomicWithRetries(t *testing.T) {
+	tests := []struct {
+		name      string
+		key       string
+		valueFunc func(oldValue interface{}) (newValue interface{}, err error)
+		setupAPI  func(api *plugintest.API)
+		wantErr   bool
+	}{
+		{
+			name: "Test SetAtomicWithRetries success after first attempt",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(true, nil)
+			},
+		},
+		{
+			name: "Test SetAtomicWithRetries success on fourth attempt",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil).Times(4)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(false, nil).Times(3)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(true, nil).Once()
+			},
+		},
+		{
+			name: "Test SetAtomicWithRetries failure on get",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				api.On("KVGet", "testNum").Return(nil, newAppError()).Once()
+			},
+			wantErr: true,
+		},
+		{
+			name: "Test SetAtomicWithRetries failure on valueFunc",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return nil, errors.New("some user provided error")
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil).Once()
+			},
+			wantErr: true,
+		},
+		{
+			name: "Test SetAtomicWithRetries DB failure on set",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil).Once()
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(false, newAppError()).Once()
+			},
+			wantErr: true,
+		},
+		{
+			name: "Test SetAtomicWithRetries failure on five set attempts -- depends on numRetries constant being = 5",
```

Comment:
Great tests!! Could I ask you to write one more where each `KVGet` Returns a different value and `KVSetWithOptions` expects a different value per iteration (with at least 3-4 iterations or maybe a max of 5).

As I think it is possible if you were to change the implementation to always pass the first value i get from `KVGet` to the `funcValue` func and have these test pass so just want to capture that case.

Reply:
> As I think it is possible if you were to change the implementation to always pass the first value i get from KVGet to the funcValue func and have these test pass so just want to capture that case.

Could you clarify this? I don't really understand the problem the test would guard against--if KVGet returned a different value each time it was called, and we give the mocked KVSetWithOptions the wrong expected OldValue, it will fail until we give it the correct OldValue. What would you be testing? Maybe I'm not understanding the example case?

Reply:
Not entirely. Basically what I want to assert is that with every iteration we assert that the value that's being passed into `valueFunc` is different and the return value from `valueFunc` that's used to invoke `KVSetWithOptions` is also different. That would assert the implementation is using the updated values for each iteration. Unless there is a better way to assert that 🤷‍♂️

Reply:
Okay, done.